File: swift/Sources/Mechbase/MeasurementPoints.swift
import Foundation

/// Create/upsert/update body for measurement points. Nil fields are omitted by the encoder.
public struct PointInput: Encodable, Sendable {
    public var name: String?
    public var externalId: String?
    public var assetId: Int?
    public var assetExternalId: String?
    public var transducerType: String?
    public var measurementUnitCode: String?
    public var location: String?
    public var bodySegment: Int?
    public var bodyAngle: Int?
    public var machineClass: String?

    public init(name: String? = nil, externalId: String? = nil, assetId: Int? = nil,
                assetExternalId: String? = nil, transducerType: String? = nil,
                measurementUnitCode: String? = nil, location: String? = nil,
                bodySegment: Int? = nil, bodyAngle: Int? = nil, machineClass: String? = nil) {
        self.name = name
        self.externalId = externalId
        self.assetId = assetId
        self.assetExternalId = assetExternalId
        self.transducerType = transducerType
        self.measurementUnitCode = measurementUnitCode
        self.location = location
        self.bodySegment = bodySegment
        self.bodyAngle = bodyAngle
        self.machineClass = machineClass
    }

    enum CodingKeys: String, CodingKey {
        case name, location
        case externalId = "external_id"
        case assetId = "asset_id"
        case assetExternalId = "asset_external_id"
        case transducerType = "transducer_type"
        case measurementUnitCode = "measurement_unit_code"
        case bodySegment = "body_segment"
        case bodyAngle = "body_angle"
        case machineClass = "machine_class"
    }
}

public struct MeasurementPoints: Sendable {
    let scope: InstallationScope

    init(http: HTTPClient, installationId: Int) {
        self.scope = InstallationScope(http: http, installationId: installationId)
    }

    public func list(
        query: String? = nil,
        assetId: Int? = nil,
        transducerType: String? = nil,
        limit: Int = 50,
        offset: Int = 0
    ) async throws -> [MeasurementPoint] {
        let q: [URLQueryItem] = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "asset_id", value: assetId.map(String.init)),
            URLQueryItem(name: "transducer_type", value: transducerType),
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "offset", value: String(offset)),
        ]
        let page = try await scope.http.get(
            scope.path("/measurement-points"),
            query: q,
            as: Page<MeasurementPoint>.self
        )
        return page.items
    }

    public func get(id: Int) async throws -> MeasurementPoint {
        try await scope.http.get(scope.path("/measurement-points/\(id)"), as: MeasurementPoint.self)
    }

    public func create(_ input: PointInput) async throws -> MeasurementPoint {
        try await scope.http.postJSON(scope.path("/measurement-points"), body: input, as: MeasurementPoint.self)
    }

    public func upsert(_ input: PointInput) async throws -> MeasurementPoint {
        try await scope.http.putJSON(scope.path("/measurement-points"), body: input, as: MeasurementPoint.self)
    }

    public func update(id: Int, _ input: PointInput) async throws -> MeasurementPoint {
        try await scope.http.patchJSON(scope.path("/measurement-points/\(id)"), body: input, as: MeasurementPoint.self)
    }

    public func delete(id: Int, cascade: Bool = false) async throws -> DeleteResult {
        let q = cascade ? [URLQueryItem(name: "cascade", value: "true")] : []
        return try await scope.http.delete(scope.path("/measurement-points/\(id)"), query: q, as: DeleteResult.self)
    }
}

File: go/pkg/mechbase/measurement_points.go
package mechbase

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
)

// MeasurementPoints is the measurement-point registry resource.
type MeasurementPoints struct {
	client *Client
	iid    int
}

// ListMeasurementPointsOptions filters the list query. Zero values are omitted.
type ListMeasurementPointsOptions struct {
	Q              string
	AssetID        int
	TransducerType string
	Limit          int
	Offset         int
}

// List returns a page of measurement points.
func (m *MeasurementPoints) List(ctx context.Context, opts ListMeasurementPointsOptions) ([]MeasurementPoint, error) {
	q := url.Values{}
	if opts.Q != "" {
		q.Set("q", opts.Q)
	}
	if opts.AssetID != 0 {
		q.Set("asset_id", fmt.Sprintf("%d", opts.AssetID))
	}
	if opts.TransducerType != "" {
		q.Set("transducer_type", opts.TransducerType)
	}
	addPaging(q, opts.Limit, opts.Offset)
	var env listEnvelope[MeasurementPoint]
	if err := m.client.doJSON(ctx, http.MethodGet, pathf(m.iid, "/measurement-points"), q, nil, &env); err != nil {
		return nil, err
	}
	return env.Items, nil
}

// Get fetches one measurement point by numeric id.
func (m *MeasurementPoints) Get(ctx context.Context, pointID int) (*MeasurementPoint, error) {
	var out MeasurementPoint
	if err := m.client.doJSON(ctx, http.MethodGet, pathf(m.iid, fmt.Sprintf("/measurement-points/%d", pointID)), nil, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// PointInput is the create/upsert/update body for measurement points.
type PointInput struct {
	Name                *string `json:"name,omitempty"`
	ExternalID          *string `json:"external_id,omitempty"`
	AssetID             *int    `json:"asset_id,omitempty"`
	AssetExternalID     *string `json:"asset_external_id,omitempty"`
	TransducerType      *string `json:"transducer_type,omitempty"`
	MeasurementUnitCode *string `json:"measurement_unit_code,omitempty"`
	Location            *string `json:"location,omitempty"`
	BodySegment         *int    `json:"body_segment,omitempty"`
	BodyAngle           *int    `json:"body_angle,omitempty"`
	MachineClass        *string `json:"machine_class,omitempty"`
}

// Create creates a measurement point.
func (m *MeasurementPoints) Create(ctx context.Context, in PointInput) (*MeasurementPoint, error) {
	var out MeasurementPoint
	if err := m.client.doJSON(ctx, http.MethodPost, pathf(m.iid, "/measurement-points"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Upsert creates or updates a measurement point, matched by external_id.
func (m *MeasurementPoints) Upsert(ctx context.Context, in PointInput) (*MeasurementPoint, error) {
	var out MeasurementPoint
	if err := m.client.doJSON(ctx, http.MethodPut, pathf(m.iid, "/measurement-points"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update partially updates a measurement point by numeric id.
func (m *MeasurementPoints) Update(ctx context.Context, pointID int, in PointInput) (*MeasurementPoint, error) {
	var out MeasurementPoint
	if err := m.client.doJSON(ctx, http.MethodPatch, pathf(m.iid, fmt.Sprintf("/measurement-points/%d", pointID)), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a measurement point; cascade also removes children.
func (m *MeasurementPoints) Delete(ctx context.Context, pointID int, cascade bool) (*DeleteResult, error) {
	q := url.Values{}
	if cascade {
		q.Set("cascade", "true")
	}
	var out DeleteResult
	if err := m.client.doJSON(ctx, http.MethodDelete, pathf(m.iid, fmt.Sprintf("/measurement-points/%d", pointID)), q, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
